hedgehog/src/Hedgehog/Internal/Property.hs

```diff
@@ -152,6 +162,55 @@ newtype PropertyT m a =
     , MonadError e
     )
 
+-- | Classifiers use strings as labels
+--
+newtype ClassifierName =
+  ClassifierName {
+      unClassifierName :: String
+    } deriving (Show, Eq, Ord)
+
+-- | A classifier can be attached to a property conditionally
+--
+--   When the amount of occurrences don't exceed the minimum percentage, a
+--   warning will be issued.
+data Classifier =
+  Classifier {
+      classifierPercentage :: !Double
+    , classifierOccurrences :: !Integer
+    } deriving (Show)
+
+-- | This semigroup is right biased, the percentage from the rightmost
+--   `Classifier` will be kept. This shouldn't be a problem since the library
+--   doesn't allow setting multiple classifiers with the same label.
+instance Semigroup Classifier where
+  (Classifier _ occ1) <> (Classifier percentage occ2) =
+    Classifier percentage (occ1 + occ2)
+
+-- | Classification are a count of how many times a property has ocurred
+--   during a test run
+--
+data Classification =
+  Classification {
+      classificationClassifiers :: !(Map ClassifierName Classifier)
+    , classificationTests :: !Integer
+    } deriving (Show)
+
+instance Semigroup Classification where
+  (Classification c1 t1) <> (Classification c2 t2) =
+    Classification
+      (Map.foldrWithKey (Map.insertWith (<>)) c1 c2)
+      (t1 + t2)
+
+instance Monoid Classification where
+  mappend =
+    (<>)
+  mempty =
+    Classification mempty 1
```

Comment:
The `1` should be `0`. You start with no tests having been run.

Reply:
If we set this to zero, we would have to increment the total on each `insertClassifier`. That will only work if the classifiers are total. If we have a set of classifiers that do not cover the complete space, then the total will be wrong. 

In other words, how would we increase the total when no classifier predicate is satisfied?

Reply:
You could also just not record the total here, as the number of tests is available in the report.

This encoding leads to an extra 1 per classification in the test suite AFAICT.

Reply:
The problem there is that a Property monad is (can be) composed of several smaller Properties - each one run as a leaf in a `TreeT`, so we need to count each node that is hit, not just the whole tree.

See the test output above, it shows it clearer than my sloppy words 😃

Reply:
That's not the purpose of the `TreeT`, or how the `PropertyT` works with it.

The rose tree in `TreeT` is used for shrinking, and yes, while `PropertyT` can be built up monadically, only the *root* (not the leaves) is traversed (or even calculated) in the happy/passing case.

When shrinking get involved, these counters will be hit in a very hard to predict manner, as the shrinking heads up through the binds (PropertyT acts sort of like a non-determinism ListT when shrinking), and would lead to completely skewed results. See #122 on that one.

But it looks like you're not using the `Classification` log from `takeSmallest` so I don't think that's a big deal (you're  `-- FIXME: maybe add classifier to failure?` should probably *not* be done).

Reply:
Yes you're right, thanks for that!

The problem still stands though, each writer will create a mempty to string along its computation. By changing the definition of mempty I don't know how to inject a count for a non-classified bind.

Reply:
Change Classification to
```haskell
newtype Classification =
  Classification {
      classificationClassifiers :: !(Map ClassifierName Classifier)
    } deriving (Show)
```

`mempty` as an empty map is fine.

And use `reportTests` to determine how good it is a render the percentage.

Reply:
The total tests in the report does not seem accurate to me - it seems to only count the root Property and not each test inside 🤔

Eg it's 1 for the classified tests above instead of 100 as it should be.

Or am I mistaken?

Reply:
Ahh, right. Your test example is a bit funky.

I would imagine most people are classifying what percentage of their generated values hits certain cases, not generating extra, deterministic values with a `for`.

I might actually go with 1 being the right answer here.

Reply:
Maybe something like this for a deterministic example:

```haskell
prop_check_classifiers :: Property
prop_check_classifiers =
  property $ do
    a <- sized $ pure . unsize
    classify (a < 50) "small number" $
      classify (a >= 50) "big number" success
```

Reply:
Sorry If I'm bring a bit brisk. BTW. I think this is really good.

Regarding the above. Essentially with that `for_` what you're doing is writing that in this single test, "I'm classifying the same state 100 times..." which is different to 100 tests being classified a single time. Or at least I get the feeling it should be.

Does that make sense?

Reply:
No worries, let's get this right 😀

In that case we should have a way to combine Classifications from different test runs, I'll have a look later this week

Reply:
No worries, let's get this right 😀

In that case we should have a way to combine Classifications from different test runs, I'll have a look later this week

Reply:
Once we have that it should be pretty easy to do the beta distribution tho 👍

Reply:
Yep, that's pretty fundamental. On line 199 of runner, it should be `classifier <> classifier0`.